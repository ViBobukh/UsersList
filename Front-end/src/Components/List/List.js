import React, {useCallback, useState} from "react";
import User from "../User/User";

/**
 * Component for create list with users
 * @param users {array}
 * @returns {JSX.Element}
 */
function List({ users }) {
  return (
    <ul>
      {users.map((user) => {
        return (
          <li key={user.id}>
            <User isNest={!!(user.items && user.items.length)} user={user} />
          </li>
        );
      })}
    </ul>
  );
}

export default List;
