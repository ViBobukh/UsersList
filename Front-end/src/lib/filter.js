/**
 * Function for filtering user by name or email
 * @param arr {array}
 * @param searchString {string}
 * @param filterType {string}
 * @returns {array}
 */
export function filter(arr, searchString, filterType) {
  const users = getUsers(arr);
  const words = searchString.split();
  return users.filter((user) => {
    return words.every((word) => {
      if (filterType === "name") {
        return user.name.toLowerCase().indexOf(word.toLowerCase()) !== -1;
      } else if (filterType === "email") {
        return user.email.toLowerCase().indexOf(word.toLowerCase()) !== -1;
      }
    });
  });
}

/**
 * Function for create array with users without nesting levels
 * @param arr {array}
 * @returns {array}
 */
export function getUsers(arr) {
  const newArr = arr.map((users) => {
    const { items, ...user } = users;
    return items && items.length ? [user, ...getUsers(items)] : [user];
  });
  return newArr.flat(1);
}
