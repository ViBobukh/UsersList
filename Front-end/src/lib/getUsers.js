/**
 * Searching current location
 * @type {any}
 */
let location = window.location;

/**
 * Get information about users from JSON
 * @returns {Promise<JSON>}
 */
export async function getUsers() {
  return await fetch(`${location}/users.json`).then((response) =>
    response.json()
  );
}
