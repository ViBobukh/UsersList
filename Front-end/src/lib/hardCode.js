export const users = [
  {
    id: 1,
    name: "Maks",
    email: "maks@gmail.com",
  },
  {
    id: 2,
    name: "Mark",
    email: "mark@gmail.com",
    items: [
      {
        id: 5,
        name: "Ivan",
        email: "ivan@gmail.com",
        items: null,
      },
      {
        id: 6,
        name: "Alex",
        email: "alex@gmail.com",
        items: [
          {
            id: 8,
            name: "Alek",
            email: "alek@gmail.com",
          },
        ],
      },
      {
        id: 7,
        name: "Denny",
        email: "denny@gmail.com",
        items: {},
      },
    ],
  },
  {
    id: 3,
    name: "Mario",
    email: "mario@gmail.com",
  },
  {
    id: 4,
    name: "Patrick",
    email: "patrick@gmail.com",
  },
];
