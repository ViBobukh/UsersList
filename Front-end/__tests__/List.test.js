import React from "react";
import renderer from "react-test-renderer";

import List from "../src/Components/List/List";
import { users } from "../src/lib/hardCode";

describe("List tests", () => {
  it("Render list", () => {
    const component = renderer.create(<List users={users} />);
    const domTree = component.toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
