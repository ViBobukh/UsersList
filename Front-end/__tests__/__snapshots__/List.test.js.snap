// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List tests Render list 1`] = `
<ul>
  <li>
    <div
      className="user"
      onClick={[Function]}
    >
      <div
        className=""
      >
        Maks
      </div>
    </div>
  </li>
  <li>
    <div
      className="user"
      onClick={[Function]}
    >
      <div
        className=""
      >
        Mark
      </div>
    </div>
  </li>
  <li>
    <div
      className="user"
      onClick={[Function]}
    >
      <div
        className=""
      >
        Mario
      </div>
    </div>
  </li>
  <li>
    <div
      className="user"
      onClick={[Function]}
    >
      <div
        className=""
      >
        Patrick
      </div>
    </div>
  </li>
</ul>
`;
